File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        tokens = s.split(" ")
        if len(pattern) != len(tokens) or len(set(pattern)) != len(set(tokens)):
            return False
        
        pattern_map = {}
        
        for i, char in enumerate(pattern):
            token = tokens[i]
            if pattern_map.setdefault(char, token) != token:
                return False
            
        return True

File: 0290-word-pattern/0290-word-pattern.cs
internal static class Extensions {
    public static IEnumerable<(int, T)> Enumerate<T>(this IEnumerable<T> input, int start = 0)
    {
        var i = start;
        foreach (var t in input) yield return (i++, t);
    }
}

public class Solution {
    public bool WordPattern(string pattern, string s)
    {
        var tokens = s.Split(' ');
        if (pattern.Length != tokens.Length || pattern.Distinct().Count() != tokens.Distinct().Count())
            return false;

        var patternMap = new Dictionary<char, string>();
        foreach (var (i, c) in pattern.Enumerate())
        {
            var token = tokens[i];
            if (patternMap.TryGetValue(c, out var storedToken))
            {
                if (token != storedToken) return false;
            }
            else
            {
                patternMap.Add(c, token);
            }
        }

        return true;
    }
}